website/guides/running-terraform-in-automation.html.md

```diff
@@ -0,0 +1,279 @@
+teams run Terraform entirely within an orchestration tool such as Jenkins.
+
+This guide covers some things that should be considered when implementing
+such automation, both to ensure safe operation of Terraform and to accommodate
+some current limitations in Terraform's workflow that require careful
+attention in automation.
+
+The guide assumes that Terraform will be running in an _non-interactive_
+environment, where it is not possible to prompt for input at the terminal.
+This is not necessarily true for wrapper scripts, but is often true when
+running in orchestration tools.
+
+This is a general guide, giving an overview of things to consider when
+implementing orchestration of Terraform. Due to its general nature, it is not
+possible to go into specifics about any particular tools, though other
+tool-specific guides may be produced later if best practices emerge around
+such a tool.
+
+## Automated Workflow Overview
+
+When running Terraform in automation, the focus is usually on the core
+plan/apply cycle. The main path, then, is the broadly same as for CLI
+usage:
+
+1. Initialize the Terraform working directory.
+2. Terraform produces a plan for changing resources to match the current
+configuration.
+3. A human review that plan, to ensure it is acceptable.
+4. Terraform takes the actions described in the plan.
+
+Steps 1, 2 and 4 can be carried out using the familiar Terraform CLI commands,
+with some additional options:
+
+* `terraform init -input=false`
+* `terraform plan -out=tfplan -input=false`
+* `terraform apply -input=false`
+
+The `-input=false` option indicates that Terraform should not attempt to
+prompt for input, and instead expect all necessary values to be provided by
+either configuration files or the command line. It may therefore be necessary
+to use the `-var` and `-var-file` options on `terraform plan` to specify any
+variable values that would traditionally have been manually-entered under
+interactive usage.
+
+## Plan and Apply on different machines
+
+When running in an orchestration tool, it can be difficult or impossible to
+ensure that the `plan` and `apply` subcommands are run on the same machine,
+in the same directory, with all of the same files present.
+
+Running `plan` and `apply` on different machines requires some additional
+steps to ensure correct behavior. A robust strategy is as follows:
+
+* After `plan` completes, archive the entire working directory, including the
+  `.terraform` subdirectory created during `init`, and save it somewhere
+  where it will be available to the apply step. A common choice is as a
+  "build artifact" within the chosen orchestration tool.
+* Before running `apply`, obtain the archive created in the previous step
+  and extract it _at the same absolute path_. This re-creates everything
+  that was present after plan, avoiding strange issues where local files
+  were created during the plan step.
```

Comment:
Should re-iterate somewhere here that the plan file can contain secrets and so should be stored securely. This is covered in the docs for `terraform plan` but worth worrying about more if the plan file is being saved as a long-lived artifact in an orchestration tool.